src/algebra/order/interval.lean

```diff
@@ -0,0 +1,336 @@
+
+instance : has_sub (interval α) := ⟨option.map₂ has_sub.sub⟩
+
+namespace nonempty_interval
+variables (s t : nonempty_interval α) (a b : α)
+
+@[simp] lemma fst_sub : (s - t).fst = s.fst - t.snd := rfl
+@[simp] lemma snd_sub : (s - t).snd = s.snd - t.fst := rfl
+@[simp] lemma coe_sub : (↑(s - t) : interval α) = s - t := rfl
+@[simp] lemma pure_sub_pure : pure a - pure b = pure (a - b) := rfl
+
+end nonempty_interval
+
+namespace interval
+variables (s t : interval α)
+
+@[simp] lemma bot_sub : ⊥ - t = ⊥ := rfl
+@[simp] lemma sub_bot : s - ⊥ = ⊥ := option.map₂_none_right _ _
+
+end interval
+end sub
+
+/-!
+### Division in ordered groups
+
+Note that this division does not apply to `ℚ` or `ℝ`.
+-/
+
+section div
+variables [preorder α] [comm_group α] [covariant_class α α (*) (≤)]
+
+@[to_additive] instance : has_div (nonempty_interval α) :=
+⟨λ s t, ⟨(s.fst / t.snd, s.snd / t.fst), div_le_div'' s.fst_le_snd t.fst_le_snd⟩⟩
+
+@[to_additive] instance : has_div (interval α) := ⟨option.map₂ (/)⟩
+
+namespace nonempty_interval
+variables (s t : nonempty_interval α) (a b : α)
+
+@[simp, to_additive] lemma fst_div : (s / t).fst = s.fst / t.snd := rfl
+@[simp, to_additive] lemma snd_div : (s / t).snd = s.snd / t.fst := rfl
+@[simp, to_additive] lemma coe_div : (↑(s / t) : interval α) = s / t := rfl
+@[simp, to_additive] lemma pure_div_pure : pure a / pure b = pure (a / b) := rfl
+
+end nonempty_interval
+
+namespace interval
+variables (s t : interval α)
+
+@[simp, to_additive] lemma bot_div : ⊥ / t = ⊥ := rfl
+@[simp, to_additive] lemma div_bot : s / ⊥ = ⊥ := option.map₂_none_right _ _
+
+end interval
+end div
+
+/-! ### Negation/inversion -/
+
+section inv
+variables [ordered_comm_group α]
+
+@[to_additive] instance : has_inv (nonempty_interval α) :=
+⟨λ s, ⟨(s.snd⁻¹, s.fst⁻¹), inv_le_inv' s.fst_le_snd⟩⟩
+
+@[to_additive] instance : has_inv (interval α) := ⟨option.map has_inv.inv⟩
+
+namespace nonempty_interval
+variables (s t : nonempty_interval α) (a : α)
+
+@[simp, to_additive] lemma fst_inv : s⁻¹.fst = s.snd⁻¹ := rfl
+@[simp, to_additive] lemma snd_inv : s⁻¹.snd = s.fst⁻¹ := rfl
+@[simp, to_additive] lemma coe_inv : (↑(s⁻¹) : interval α) = s⁻¹ := rfl
+@[simp, to_additive] lemma inv_pure : (pure a)⁻¹ = pure a⁻¹ := rfl
+
+end nonempty_interval
+
+@[simp, to_additive] lemma interval.inv_bot : (⊥ : interval α)⁻¹ = ⊥ := rfl
+
+end inv
+
+namespace nonempty_interval
+variables [ordered_comm_group α] {s t : nonempty_interval α}
+
+@[to_additive] protected lemma mul_eq_one_iff :
+  s * t = 1 ↔ ∃ a b, s = pure a ∧ t = pure b ∧ a * b = 1 :=
+begin
+  refine ⟨λ h, _, _⟩,
+  { rw [ext_iff, prod.ext_iff] at h,
+    have := (mul_le_mul_iff_of_ge s.fst_le_snd t.fst_le_snd).1 (h.2.trans h.1.symm).le,
+    refine ⟨s.fst, t.fst, _, _, h.1⟩; ext; try { refl },
+    exacts [this.1.symm, this.2.symm] },
+  { rintro ⟨b, c, rfl, rfl, h⟩,
+    rw [pure_mul_pure, h, pure_one] }
+end
+
+-- @[to_additive]
```

Comment:
Btw, the reason we need this duplication is because `interval.has_div` (purposefully) doesn't match `interval.has_sub`. I guess if we had `has_ordered_div`, the problem would also be fixed, but the consensus is that we don't need it.